activesupport/lib/active_support/callbacks.rb

```diff
@@ -349,23 +334,96 @@ def duplicates?(other)
         # Wraps code with filter
         def apply(callback_sequence)
           user_conditions = conditions_lambdas
-          user_callback = make_lambda @filter
+          user_callback = CallTemplate.build(@filter, self)
 
           case kind
           when :before
-            Filters::Before.build(callback_sequence, user_callback, user_conditions, chain_config, @filter)
+            Filters::Before.build(callback_sequence, user_callback.make_lambda, user_conditions, chain_config, @filter)
           when :after
-            Filters::After.build(callback_sequence, user_callback, user_conditions, chain_config)
+            Filters::After.build(callback_sequence, user_callback.make_lambda, user_conditions, chain_config)
           when :around
-            Filters::Around.build(callback_sequence, user_callback, user_conditions, chain_config)
+            callback_sequence.around(user_callback, user_conditions)
           end
         end
 
+        def current_scopes
+          Array(chain_config[:scope]).map { |s| public_send(s) }
+        end
+
         private
+          def compute_identifier(filter)
+            case filter
+            when String, ::Proc
+              filter.object_id
+            else
+              filter
+            end
+          end
 
-          def invert_lambda(l)
-            lambda { |*args, &blk| !l.call(*args, &blk) }
+          def conditions_lambdas
+            @if.map { |c| CallTemplate.build(c, self).make_lambda } +
+              @unless.map { |c| CallTemplate.build(c, self).inverted_lambda }
           end
+      end
+
+      # A future invocation of user-supplied code (either as a callback,
+      # or a condition filter).
+      class CallTemplate # :nodoc:
+        def initialize(target, method, arguments, block)
+          @override_target = target
+          @method_name = method
+          @arguments = arguments
+          @override_block = block
+        end
+
+        # Return the parts needed to make this call, with the given
+        # input values.
+        #
+        # Returns an array of the form:
+        #
+        #   [target, block, method, *arguments]
+        #
+        # This array can be used as such:
+        #
+        #   target.send(method, *arguments, &block)
+        #
+        # The actual invocation is left up to the caller to minimize
+        # call stack pollution.
+        def expand(target, value, block)
+          result = @arguments.map { |arg|
+            case arg
+            when :value; value
+            when :target; target
+            when :block; block or raise ArgumentError
+            end
+          }
+
+          result.unshift @method_name
+          result.unshift @override_block || block
+          result.unshift @override_target || target
```

Comment:
This awkward construction of the array keeps us down to one allocation, in the map. The callers likewise shift the values back out, for the same reason.